src/endpoint_memory.rs

```diff
@@ -1,140 +1,230 @@
+    Word16x1,
+    /// 16x2 bits per word. Each 32-bit word is accessed as two 16-bit half-words.
+    ///
+    /// This matches the behavior of `EP_MEMORY_ACCESS_2X16 = true` from previous versions of this library.
+    Word16x2,
+    /// 32x1 bits per word. Each 32-bit word is accessed as one 32-bit word.
+    Word32x1,
+}
+
+impl MemoryAccess {
+    /// Value to multiply offsets within the EP memory by when calculating address to read or write to.
+    #[inline(always)]
+    const fn offset_multiplier(self) -> usize {
+        match self {
+            MemoryAccess::Word16x1 => 2,
+            MemoryAccess::Word16x2 | MemoryAccess::Word32x1 => 1,
+        }
+    }
+
+    /// Size of unit used when reading and writing EP memory, in bytes.
+    #[inline(always)]
+    const fn unit_size(self) -> usize {
+        match self {
+            MemoryAccess::Word16x1 | MemoryAccess::Word16x2 => 2,
+            MemoryAccess::Word32x1 => 4,
+        }
+    }
+}
+
 pub struct EndpointBuffer<USB> {
-    mem: &'static mut [VolatileCell<u16>],
+    mem_ptr: *mut (),
+    mem_len: usize,
     marker: PhantomData<USB>,
 }
 
+unsafe impl<USB> Send for EndpointBuffer<USB> {}
+
 impl<USB: UsbPeripheral> EndpointBuffer<USB> {
     pub fn new(offset_bytes: usize, size_bytes: usize) -> Self {
-        let ep_mem_ptr = USB::EP_MEMORY as *mut VolatileCell<u16>;
-
-        let offset_words = offset_bytes >> 1;
-        let count_words = size_bytes >> 1;
-        let offset_u16_words;
-        let count_u16_words;
-        if USB::EP_MEMORY_ACCESS_2X16 {
-            offset_u16_words = offset_words;
-            count_u16_words = count_words;
-        } else {
-            offset_u16_words = offset_words * 2;
-            count_u16_words = count_words * 2;
-        };
-
-        unsafe {
-            let mem = slice::from_raw_parts_mut(ep_mem_ptr.add(offset_u16_words), count_u16_words);
-            Self {
-                mem,
-                marker: PhantomData,
-            }
-        }
-    }
+        let mem_offset_bytes = offset_bytes * USB::EP_MEMORY_ACCESS.offset_multiplier();
+        let mem_len = size_bytes * USB::EP_MEMORY_ACCESS.offset_multiplier() / USB::EP_MEMORY_ACCESS.unit_size();
 
-    #[inline(always)]
-    fn read_word(&self, index: usize) -> u16 {
-        if USB::EP_MEMORY_ACCESS_2X16 {
-            self.mem[index].get()
-        } else {
-            self.mem[index * 2].get()
+        let mem_ptr = unsafe { USB::EP_MEMORY.byte_add(mem_offset_bytes).cast_mut() };
+        Self {
+            mem_ptr,
+            mem_len,
+            marker: PhantomData,
         }
     }
 
     #[inline(always)]
-    fn write_word(&self, index: usize, value: u16) {
-        if USB::EP_MEMORY_ACCESS_2X16 {
-            self.mem[index].set(value);
-        } else {
-            self.mem[index * 2].set(value);
-        }
+    fn get_mem_slice<T>(&self) -> &mut [VolatileCell<T>] {
```

Comment:
I believe this function is very unsafe because calling it twice withing the same block is unsound.

Reply:
That's fair. I realized that I can return `&[VolatileCell<T>]`, which relaxes the safety requirements quite a bit. I did also mark it as unsafe for posterity.